### PR TITLE
Add Heap and Resizable arrays

### PR DESCRIPTION
**Pull request by acorrenson**
I added simple Heap (imperative implementation) and Resizable arrays (since they are required to implement Heaps)

### CODE REVIEW

**Comment by imartayan** [comment]
I am not entirely sure about this, but I think that in the `sift_up` function, the parent index should be (i-1)/2 and not i/2.

**Comment by acorrenson** [comment]
> I am not entirely sure about this, but I think that in the `sift_up` function, the parent index should be (i-1)/2 and not i/2.

MMh, i'll check

**Comment by acorrenson** [comment]
Damn, you're right ^^'. My bad.... It's a mess since we need to remap the indices (classical formulas on Heap work with indices starting at 1)

**Comment by acorrenson** [comment]
I fixed it in most of the functions, but i forgot to fix "sift_up"...
Maybe we should waste the first cell of the Heap to apply standard formulas without having to remap everything... What do u think ?

**Comment by imartayan** [comment]
Maybe... The first cell could also be used to store the size, if we do not want to waste it.

**Comment by Moxinilian** [comment]
By doing that you would kill the generics tho.

**Comment by acorrenson** [comment]
I thought the exact same thing before realizing that we want to keep the polymorphism... (It's frustrating since in memory, everything is an integer in ocaml)

**Comment by acorrenson** [comment]
Anyway, I updated the code for a safer index system

**Review by imartayan — APPROVED** [review]
Seems good to me.

File: OCaml_implem/heap/heap.ml [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    c := parent !c;
+  done
+
+let insert h v =
+  set (h.values) (h.size) v;
+  sift_up h h.size;
+  h.size <- h.size + 1
+
+let to_array h =
+  let open Resizable_array in
```
Comment (Drakzia) [inline OCaml_implem/heap/heap.ml]: Useless line ? (open Resizable_array is done in the beginning of the file)

**Review by Drakzia — APPROVED** [review]
LGTM appart from line 56 in heap.ml -> see comment

**Pull request merged**